### PR TITLE
lpython: new port in lang

### PR DESCRIPTION
**Pull request by barracuda156**
#### Description

New port https://lpython.org

###### Type(s)

- [ ] bugfix
- [ ] enhancement
- [ ] security fix

###### Tested on

macOS 14.2.1
Xcode 15.1

macOS 10.6
Xcode 3.2

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?
- [ ] checked that the Portfile's most important [variants](https://trac.macports.org/wiki/Variants) haven't been broken?

### CODE REVIEW

File: lang/lpython/Portfile [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+PortGroup               compiler_blacklist_versions 1.0
+PortGroup               github 1.0
+
+github.setup            lcompilers lpython 0.20.0 v
+revision                0
+categories              lang python
+maintainers             {@barracuda156 gmail.com:vital.had} openmaintainer
+license                 BSD
+description             High-performance typed Python compiler
+long_description        LPython aggressively optimizes type-annotated Python code. \
```
Comment (pmetzger) [inline lang/lpython/Portfile:14]: Thanks for working on this.

84 characters > 80 characters. Please do clean this up.

File: lang/lpython/Portfile [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+                        ${prefix}/lib/${port_libfmt}/cmake
+
+depends_build-append    port:bash \
+                        port:bison \
+                        port:re2c
+depends_lib-append      path:lib/pkgconfig/RapidJSON.pc:rapidjson \
+                        port:zlib \
+                        port:${port_libfmt}
+
+set py_ver              3.11
```
Comment (pmetzger) [inline lang/lpython/Portfile:35]: Does this actually require 3.11? Our default now is 3.12.

File: lang/lpython/files/0003-Install-libs-into-a-sane-location.patch [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+--- src/bin/CMakeLists.txt
++++ src/bin/CMakeLists.txt
+@@ -73,8 +73,8 @@ endif()
+ 
+ install(TARGETS lpython
+     RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+-    ARCHIVE DESTINATION share/lpython/lib
+-    LIBRARY DESTINATION share/lpython/lib
++    ARCHIVE DESTINATION lib
++    LIBRARY DESTINATION lib
```
Comment (pmetzger) [inline lang/lpython/files/0003-Install-libs-into-a-sane-location.patch:21]: Might a subdir of lib make more sense? Share certainly doesn't make sense of course.

**Comment by pmetzger** [comment]
I have some comments above. Thanks for working on this. Would like an opinion from @reneeotten as he's the python domain expert.

File: lang/lpython/Portfile [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                        -DWITH_RUNTIME_LIBRARY=YES \
+                        -DWITH_STACKTRACE=NO \
+                        -DWITH_UNWIND=NO \
+                        -DWITH_ZLIB=YES
+
+pre-configure {
+    system -W ${worksrcpath} "sh ./build0.sh"
+}
+
+cmake.build_type        Release
```
Comment (reneeotten) [inline lang/lpython/Portfile:76]: can this be left to the default, which is (I think) "macports" and sets correct flags and such?

File: lang/lpython/files/0002-UserOverride.cmake-fix-native-optflag-for-PowerPC.patch [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+ if (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+     # g++
+     set(common "-Wall -Wextra")
+-    set(CMAKE_CXX_FLAGS_RELEASE_INIT "${common} -O3 -march=native -funroll-loops -DNDEBUG")
++    if (CMAKE_SYSTEM_PROCESSOR MATCHES "ppc|powerpc")
++        set(native "-mtune=native")
++    else ()
++        set(native "-march=native")
++    endif ()
++    set(CMAKE_CXX_FLAGS_RELEASE_INIT "${common} -O3 ${native} -funroll-loops -DNDEBUG")
```
Comment (reneeotten) [inline lang/lpython/files/0002-UserOverride.cmake-fix-native-optflag-for-PowerPC.patch:24]: I didn't look at the full CMake file but nothing should be compiled with `native`; that should all be stripped and only be done in a `native` variant. Similarly the correct flag is `-Os` not `-O3`

**Comment by barracuda156** [comment]
@pmetzger @reneeotten I will address these tonight.

File: lang/lpython/files/0003-Install-libs-into-a-sane-location.patch (hunk 3 above)
Comment (barracuda156) [reply lang/lpython/files/0003-Install-libs-into-a-sane-location.patch:21]: There are just two libraries being installed. I do not see any point in a subdirectory, to be honest, and then a need to pass a flag every time to use it.

File: lang/lpython/Portfile (hunk 2 above)
Comment (barracuda156) [reply lang/lpython/Portfile:35]: Until/unless a solution is found to have 3.12 fixed for PPC, I am not using it is any of my ports. (Of course, if you rather not have the port available at all, PR can be closed. But there are multiple existing ports using Python 3.11 and even earlier as the default choice.)

File: lang/lpython/files/0002-UserOverride.cmake-fix-native-optflag-for-PowerPC.patch (hunk 5 above)
Comment (barracuda156) [reply lang/lpython/files/0002-UserOverride.cmake-fix-native-optflag-for-PowerPC.patch:24]: Not really sure about optimization level (it is not my choice but of the upstream), but I can change it.

**Comment by barracuda156** [comment]
@reneeotten Will you be interested to maintain this port by the way? I do not see much point to add myself at present stage.

File: lang/lpython/Portfile (hunk 2 above)
Comment (pmetzger) [reply lang/lpython/Portfile:35]: 3.12 is our default right now. If you don't want to work on it, of course, that's fine. It's a volunteer project.

Comment (barracuda156) [reply lang/lpython/Portfile:35]: Okay, sure. Anyone else is free to borrow this and submit the port in any form desirable. No objections from my side.

**Comment by barracuda156** [comment]
@reneeotten Since it may not be clear, my only issue here was being demanded to use a broken Python 3.12. Everything else could have been addressed.

If you will be interested to have this port under your maintenance, please feel free to borrow the portfile.

**Pull request closed without merging**